lib/spack/docs/tutorial_advanced_packaging.rst

```diff
@@ -0,0 +1,552 @@
+
+  root@advanced-packaging-tutorial:/# spack install netcdf
+  ==> libsigsegv is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/libsigsegv-2.11-fypapcprssrj3nstp6njprskeyynsgaz
+  ==> m4 is already installed in /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/m4-1.4.18-r5envx3kqctwwflhd4qax4ahqtt6x43a
+  ...
+  ==> Installing netcdf
+  ==> Using cached archive: /usr/local/var/spack/cache/netcdf/netcdf-4.4.1.1.tar.gz
+  ==> Already staged netcdf-4.4.1.1-gk2xxhbqijnrdwicawawcll4t3c7dvoj in /usr/local/var/spack/stage/netcdf-4.4.1.1-gk2xxhbqijnrdwicawawcll4t3c7dvoj
+  ==> Already patched netcdf
+  ==> Building netcdf [AutotoolsPackage]
+  ==> Executing phase: 'autoreconf'
+  ==> Executing phase: 'configure'
+  ==> Executing phase: 'build'
+  ==> Executing phase: 'install'
+  ==> Successfully installed netcdf
+    Fetch: 0.01s.  Build: 24.61s.  Total: 24.62s.
+  [+] /usr/local/opt/spack/linux-ubuntu16.04-x86_64/gcc-5.4.0/netcdf-4.4.1.1-gk2xxhbqijnrdwicawawcll4t3c7dvoj
+
+
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Single package providing multiple virtual specs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+At the close of this tutorial's subsection, it may be useful to see where the
+build-interface protocol shines the most i.e. when it comes to manage packages
+that provide more than one virtual spec. An example of a package of this kind is
+``intel-parallel-studio``, and due to its complexity we'll limit our discussion
+here to just a few considerations (without any hands-on). You can open
+the related ``package.py`` in the usual way:
+
+.. code-block:: console
+
+  root@advanced-packaging-tutorial:/# spack edit intel-parallel-studio
+
+As you can see this package provides a lot of virtual specs, and thus it has
+more than one function that enters into the build-interface protocol. These
+functions will be invoked for *exactly the same spec* according to the key used
+by its dependents in the subscript query.
+
+So, for instance, the ``blas_libs`` property will be returned when
+``intel-parallel-studio`` is the BLAS provider in the current DAG and
+is retrieved by a dependent with:
+
+.. code-block:: python
+
+  blas = self.spec['blas']
+  blas_libs = blas.libs
+
+Within the property we inspect various aspects of the current spec:
+
+.. code-block:: python
+
+  @property
+  def blas_libs(self):
+     spec = self.spec
+     prefix = self.prefix
+     shared = '+shared' in spec
+
+     if '+ilp64' in spec:
+         mkl_integer = ['libmkl_intel_ilp64']
+     else:
+         mkl_integer = ['libmkl_intel_lp64']
+     ...
+
+and construct the list of library we need to return accordingly.
+
+What we achieved is that the complexity of dealing with ``intel-parallel-studio``
+is now gathered in the package itself, instead of being spread
+all over its possible dependents. A package using MPI, or LAPACK, thus don't care
+anymore what is the actual implementation that is going to be used, and has
```

Comment:
This sentence is worded a little funny. I would say: "Thus, a package that uses MPI or LAPACK doesn't care which implementation it uses, as each virtual dependency has *a uniform interface* ..."

This is definitely an important point, and I'm not sure if you're doing enough to drive the point home. I would go further and say something like "without this interface" or "before this interface was added", "each package that depended on MPI or LAPACK had dozens of lines of code copied from other packages telling it where to find the libraries and what they are called. With the addition of this interface, the virtual dependency itself tells other packages that depend on it where it can find its libraries."